backend/src/controller/chatController.ts

```diff
@@ -2,136 +2,199 @@ import { Response } from 'express';
-		req.session.levelState[currentLevel].chatHistory.push({
-			completion: null,
-			chatMessageType: CHAT_MESSAGE_TYPE.USER,
-			infoMessage: message,
-		});
-	}
+		transformedMessageCombined ?? null
+	).reduce(pushMessageToHistory, chatHistory);
 
 	// detect defences on input message
-	const triggeredDefencesPromise = detectTriggeredDefences(
+	const triggeredInputDefencesPromise = detectTriggeredInputDefences(
 		message,
-		req.session.levelState[currentLevel].defences
-	).then((DefenceReport) => {
-		chatResponse.defenceReport = DefenceReport;
-	});
+		defences
+	);
 
 	// get the chatGPT reply
 	const openAiReplyPromise = chatGptSendMessage(
-		req.session.levelState[currentLevel].chatHistory,
-		req.session.levelState[currentLevel].defences,
+		chatHistoryWithNewUserMessages,
+		defences,
 		chatModel,
-		transformedMessage
-			? combineTransformedMessage(transformedMessage)
-			: message,
-		transformedMessage ? true : false,
-		req.session.levelState[currentLevel].sentEmails,
+		transformedMessageCombined ?? message,
 		currentLevel
 	);
 
 	// run defence detection and chatGPT concurrently
-	const [, openAiReply] = await Promise.all([
-		triggeredDefencesPromise,
+	const [inputDefenceReport, openAiReply] = await Promise.all([
+		triggeredInputDefencesPromise,
 		openAiReplyPromise,
 	]);
 
-	// if input message is blocked, restore the original chat history and add user message (not as completion)
-	if (chatResponse.defenceReport.isBlocked) {
-		// restore the original chat history
-		req.session.levelState[currentLevel].chatHistory = chatHistoryBefore;
+	const botReply = openAiReply.chatResponse.completion?.content?.toString();
+	const outputDefenceReport = botReply
+		? detectTriggeredOutputDefences(botReply, defences)
+		: null;
 
-		req.session.levelState[currentLevel].chatHistory.push({
-			completion: null,
-			chatMessageType: CHAT_MESSAGE_TYPE.USER,
-			infoMessage: message,
-		});
-	} else {
-		chatResponse.wonLevel = openAiReply.wonLevel;
-		chatResponse.reply = openAiReply.completion?.content?.toString() ?? '';
+	const defenceReports = outputDefenceReport
+		? [inputDefenceReport, outputDefenceReport]
+		: [inputDefenceReport];
+	const combinedDefenceReport = combineChatDefenceReports(defenceReports);
 
-		// combine triggered defences
-		chatResponse.defenceReport.triggeredDefences = [
-			...chatResponse.defenceReport.triggeredDefences,
-			...openAiReply.defenceReport.triggeredDefences,
-		];
-		// combine blocked
-		chatResponse.defenceReport.isBlocked = openAiReply.defenceReport.isBlocked;
+	// if blocked, restore original chat history and add user message to chat history without completion
+	const updatedChatHistory = combinedDefenceReport.isBlocked
+		? pushMessageToHistory(chatHistory, {
+				completion: null,
+				chatMessageType: CHAT_MESSAGE_TYPE.USER,
+				infoMessage: message,
+		  })
+		: openAiReply.chatHistory;
```

Comment:
I smelled a regression here and did an investigation:

regression:
Suppose the user sends a message in a higher level and the input is blocked, but also when we call getFinalReplyAfterAllToolCalls to get chatGPTReply something goes wrong, and we have an empty response.
Before:
[problem caught] return early out of chatGPTSendMessage, but pass on values for completion, defenceReport, wonLevel and openAIErrorMessage
[input blocked] restore original chat history and mutate history to add user message as info. Slice emails, send response
else set chatReponse defenceReport according to output defences
[output blocked (won't happen because output is "")] (we don't restore history!) mustate history to add block message. Slice emails, send response.
[problem caught] add error message to history and send error response
Now:
continue with new sent emails, wonLevel, openAIErrorMessage, original history
[blocked, but only by input defence since reply is ""] restore original chat history and add user message as infoMessage. Bot reply is set to empty
[problem] mutate chat history to add error, then send error response
update history to add blocked message
mutate state with updated story and sent emails and return response
To conclude:
both before and now, if input message is blocked (by input defences) but also we get an error when getting a reply, the blocking takes precidence over the error and we respond with a "you've been blocked" kind of message ✅
both before and now, if input message not blocked but also we get an error when getting a reply, we won't trigger any output defences, and so we add an error message to history and return an error response ✅
if input is not blocked, we successfully get a reply and the output is blocked, before we would add the block message to the already-mutated history. Now we restore the original history, then add user message as info and then add the block message ❌

Reply:
see discussion [here](https://github.com/ScottLogic/prompt-injection/pull/743#discussion_r1458894955)